src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -1367,6 +1367,11 @@ HasRecurringTuples(Node *node, RecurringTuplesType *recurType)
 			return true;
 		}
 
+		if (FindNodeCheckInRangeTableList(query->rtable, IsDistributedTableRTE))
```

Comment:
I like how small this change is, but this function is becoming a little odd. If I look at the `if (IsA(node, RangeTblEntry))` block it appears as though we return true if any RTE is a reference table. In reality, we already look at all the RTEs here so we would return false if there is a reference table and a distributed table. The function would also behave quite differently if we passed an RTE or rtable directly. It would help to do some refactoring / write this function from scratch.

Reply:
Here is my take: I think with this change, we're actually breaking `HasRecurringTuples()` function.

I think the right way to do is refactor `DeferErrorIfFromClauseRecurs()` and its callers such that `if (!queryTree->hasSubLinks)` is not there anymore. And, finally, call `DeferErrorIfFromClauseRecurs()` in `DeferErrorIfUnsupportedUnionQuery()`  instead of `HasRecurringTuples()`.

 `DeferErrorIfFromClauseRecurs()` is already broken (e.g., it doesn't do what its name suggests due to the `if` statement) and if we merge this, we'd break `HasRecurringTuples()`.

I think refactoring  is straight forward, simply pull  `if (!queryTree->hasSubLinks)` to the places where `DeferErrorIfFromClauseRecurs()`. 

Does that make sense to you as well?

Reply:
Complication of what we are trying to do is causing the issue
We want HasRecurringTuples to return
1 - true : if the query only has a reference table
2 - true : if the query does not reference any table (function, values etc)
3 - false : if the query references at least one distributed table.

Originally, I want to restrict the function such that it will check if the reference table is involved in a join with a distributed table. But we decided to go with presence of a distributed table instead.

`HasRecurringTuples` is called from 3 external places. 1 call from `IsRecurringRangeTable` and 2 calls from  `DeferErrorIfUnsupportedUnionQuery`. Our starting point was to fix the second call.

`IsRecurringRangeTable` is indirectly called from many other places through `RelationInfoContainsOnlyRecurringTuples` and  `RelationInfoContainsRecurringTuples`. Thus, even though the changes are small, it is affects more places it is intended to do.

Given that, it would have been nicer limiting the change with subquery union pushdown.

@onderkalaci I am not following comment. 

> Here is my take: I think with this change, we're actually breaking HasRecurringTuples() function.

Function was too restrictive. It had (still has) many false positives. This change removes some restrictions and allows some false negatives relying that these would be caught later by join checks.

Let me spend some time on your suggestion on not using HasRecurringTuples.

Reply:
@onderkalaci There are 2 issues with that approach
1 - `DeferErrorIfFromClauseRecurs` only checks distributed table presence in immediate range table entries. Thus if union includes a join wrapped inside another subquery, we won't be able to detect that. This would cause future regressions on this matter.

2 - Less of an issue, we would need to do some more refactoring to return correct error message.